### ISSUE TITLE
function to increase delegator stake?

### ISSUE THREAD

**Issue by MKong** [Fantom-foundation/fantom-sfc#6]
Is there a function to increase delegator stake?

**Comment by devintegral3** [comment]
No, unfortunately this method cannot be added without increasing complexity. Simply said, contract won't know which delegation amount should be used (while calculation rewards for an epoch) if it's changeable. It's possible to add it, but contract will become more complex. On another hand, user may simply create a new delegation, because minimum delegation is only 1 FTM

**Comment by rekt0x** [comment]
@devintegral3 I guess the same applies for a function to re-delegate FTM, right? It would be nice, so delegators can change their validator without losing rewards. But I believe that would cause the same issue because the contract won't know how to calculate rewards for the epoch because delegations could be changed.

**Comment by rekt0x** [comment]
@devintegral3 but isn't it the same with the `increaseStake()` functions for validators? why can a validator increase its stake but delegators can not? this should affect the reward calculation in the same way in my opinion.

**Comment by rekt0x** [comment]
It would be great to see following functions getting added:
- increase delegation
- partially withdraw delegation
- partially withdraw stake

**Comment by devintegral3** [comment]
> but isn't it the same with the increaseStake() functions for validators? why can a validator increase its stake but delegators can not? this should affect the reward calculation in the same way in my opinion.

Because validator's stakers are being copied into each epoch snapshot. During rewards calculation, contract doesn't use current validator's stake, but the one which is written into epoch snapshot for a specified epoch. It's needed to allow delegators to claim their rewards even if validator has withdrawn (also it allows to implement increaseStake).

> It would be great to see following functions getting added:
> * increase delegation
> * partially withdraw delegation
> * partially withdraw stake

Thank you for your feedback. Those features can be added, with the following limitations:

`increaseDelegation` can be implemented if this call will require to claim all the previous rewards before calling this function. This way, we can avoid storing a history of delegation changes.

It'll require a hardfork because we'll have to add support of new log types into go-lachesis node (support of UpdatedDelegation and UpdatedStake was added in https://github.com/Fantom-foundation/go-lachesis/pull/414 ). All the stakes must be tracked by go-lachesis node using the special-purpose contract logs. Instead of adding support for a specific log like `IncreasedDelegation`, we add the general-purpose log which could be used for other features as well. We can emit `IncreasedDelegation` as well for clarity, but only `UpdatedDelegation/UpdatedStake` will be tracked by go-lachesis node.

Using the same new types of logs, support of partial withdrawal can be added as well.

**Comment by rekt0x** [comment]
> Because validator's stakes are being copied into each epoch snapshot. During rewards calculation, contract doesn't use current validator's stake, but the one which is written into epoch snapshot for a specified epoch. It's needed to allow delegators to claim their rewards even if validator has withdrawn (also it allows to implement increaseStake).
ok, makes sense.

>`increaseDelegation` can be implemented if this call will require to claim all the previous rewards before calling this function. This way, we can avoid storing a history of delegation changes.

Can you make that work in combination with the reward lock? I think this implementation would be fine, but people should not lose their rewards in case they update their delegation. If they withdraw, then they should, but not if they increase their delegation or also re-delegate to another validator.

>... we add the general-purpose log which could be used for other features as well.
sounds good 👌

**Comment by devintegral3** [comment]
> Can you make that work in combination with the reward lock? I think this implementation would be fine, but people should not lose their rewards in case they update their delegation. If they withdraw, then they should, but not if they increase their delegation or also re-delegate to another validator.

Certainly delegators shouldn't lose their rewards during increasing delegation.

But it has to be discussed how exactly we change rewards burning logic - #17
